src/app/Rules/CreateRule.tsx

```diff
@@ -286,26 +286,20 @@ const CreateRuleForm: React.FC<CreateRuleFormProps> = ({ ...props }) => {
   }, [addSubscription, context.targets, setTargets]);
 
   React.useEffect(() => {
-    // Set validations
-    let validation: ValidatedOptions = ValidatedOptions.default;
-    let matches: Target[] = [];
     if (matchExpression !== '' && targets.length > 0) {
-      try {
-        matches = targets.filter((t) => {
-          const res = evaluateTargetWithExpr(t, matchExpression);
-          if (typeof res === 'boolean') {
-            return res;
-          }
-          throw new Error('The expression matching failed.');
-        });
-        validation = matches.length ? ValidatedOptions.success : ValidatedOptions.warning;
-      } catch (err) {
-        validation = ValidatedOptions.error;
-      }
+      addSubscription(
+        context.api.matchTargetsWithExpr(matchExpression, targets).subscribe({
+          next: (ts) => {
+            setMatchExpressionValid(ts.length ? ValidatedOptions.success : ValidatedOptions.warning);
+            matchedTargets.next(ts);
+          },
+          error: (_) => {
+            setMatchExpressionValid(ValidatedOptions.error);
+          },
+        })
+      );
     }
-    setMatchExpressionValid(validation);
-    matchedTargets.next(matches);
-  }, [matchExpression, targets, matchedTargets, setMatchExpressionValid]);
+  }, [matchExpression, targets, matchedTargets, context.api, setMatchExpressionValid, addSubscription]);
```

Comment:
This same comment applies to multiple similar changes following the same pattern -

This looks like it will fire an API request on every keystroke updating the matchExpression. If that's the case, I think there should be a `debounceTime` or similar RxJS operator applied so that the request is only fired after the user has stopped typing for some time.

Reply:
Oh yesss! There is actually a match expression hook with `debountime`. I will update all components to use this instead.

Reply:
Updated now. Tho, some will use the expression service directly so avoid calling unnecessary effect hook. Tho, we will need to use `switchMap` to ignore previous requests too.

Need a little more refactoring :))